File: packages/go/parameters.go
package openrouterauto

import (
	"encoding/json"
	"fmt"

	_ "embed"
)

//go:embed registry/parameters.json
var parametersJSON []byte

//go:embed registry/platform-params.json
var platformParamsJSON []byte

type parameterDef struct {
	Type    string   `json:"type"`
	Min     *float64 `json:"min"`
	Max     *float64 `json:"max"`
	Default any      `json:"default"`
	Enum    []any    `json:"enum"`
}

var (
	paramDefs      map[string]parameterDef
	platformParams map[string]bool
)

func init() {
	if err := json.Unmarshal(parametersJSON, &paramDefs); err != nil {
		panic("openrouterauto: failed to load parameters.json: " + err.Error())
	}
	var list []string
	if err := json.Unmarshal(platformParamsJSON, &list); err != nil {
		panic("openrouterauto: failed to load platform-params.json: " + err.Error())
	}
	platformParams = make(map[string]bool, len(list))
	for _, p := range list {
		platformParams[p] = true
	}
}

// ValidateParameters checks params against the registry definitions for model.
// Returns a slice of human-readable error strings (empty = valid).
func ValidateParameters(model Model, params map[string]any) []string {
	supported := make(map[string]bool, len(model.SupportedParameters))
	for _, p := range model.SupportedParameters {
		supported[p] = true
	}
	var errsOut []string
	for name, value := range params {
		// Platform-level params are always allowed.
		if platformParams[name] {
			continue
		}
		if !supported[name] {
			errsOut = append(errsOut, fmt.Sprintf("%s is not supported by model %s", name, model.ID))
			continue
		}
		def, ok := paramDefs[name]
		if !ok {
			continue // unknown param — pass through
		}
		if msg := validateParamValue(name, value, def); msg != "" {
			errsOut = append(errsOut, msg)
		}
	}
	return errsOut
}

// GetDefaultParameters returns a map of param name → default value for a model.
func GetDefaultParameters(model Model) map[string]any {
	defaults := make(map[string]any)
	for _, name := range model.SupportedParameters {
		if def, ok := paramDefs[name]; ok && def.Default != nil {
			defaults[name] = def.Default
		}
	}
	return defaults
}

func validateParamValue(name string, value any, def parameterDef) string {
	switch def.Type {
	case "number":
		v, ok := toFloat64(value)
		if !ok {
			return fmt.Sprintf("%s must be a number", name)
		}
		if def.Min != nil && v < *def.Min {
			return fmt.Sprintf("%s must be >= %g", name, *def.Min)
		}
		if def.Max != nil && v > *def.Max {
			return fmt.Sprintf("%s must be <= %g", name, *def.Max)
		}
	case "integer":
		// JSON numbers decode as float64; accept both float64 (whole) and int.
		switch v := value.(type) {
		case float64:
			if v != float64(int(v)) {
				return fmt.Sprintf("%s must be an integer", name)
			}
		case int:
			// ok
		default:
			return fmt.Sprintf("%s must be an integer", name)
		}
	case "boolean":
		if _, ok := value.(bool); !ok {
			return fmt.Sprintf("%s must be a boolean", name)
		}
	case "string":
		if _, ok := value.(string); !ok {
			return fmt.Sprintf("%s must be a string", name)
		}
	}
	return ""
}

func toFloat64(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	}
	return 0, false
}

File: packages/rust/src/parameters.rs
use crate::types::Model;
use once_cell::sync::Lazy;
use serde::Deserialize;
use std::collections::{HashMap, HashSet};

const PARAMETERS_JSON: &str = include_str!("../registry/parameters.json");
const PLATFORM_PARAMS_JSON: &str = include_str!("../registry/platform-params.json");

#[derive(Debug, Deserialize)]
struct ParamDef {
    #[serde(rename = "type")]
    param_type: String,
    #[serde(default)]
    min: Option<f64>,
    #[serde(default)]
    max: Option<f64>,
    #[serde(default)]
    default: Option<serde_json::Value>,
}

static PARAM_DEFS: Lazy<HashMap<String, ParamDef>> =
    Lazy::new(|| serde_json::from_str(PARAMETERS_JSON).expect("failed to parse parameters.json"));

static PLATFORM_PARAMS: Lazy<HashSet<String>> = Lazy::new(|| {
    let list: Vec<String> =
        serde_json::from_str(PLATFORM_PARAMS_JSON).expect("failed to parse platform-params.json");
    list.into_iter().collect()
});

/// Validate parameters against the registry definitions for a model.
/// Returns a list of human-readable error strings (empty = valid).
pub fn validate_parameters(
    model: &Model,
    params: &HashMap<String, serde_json::Value>,
) -> Vec<String> {
    let supported: HashSet<&str> = model.supported_parameters.iter().map(|s| s.as_str()).collect();
    let mut errors = Vec::new();

    for (name, value) in params {
        // Platform-level params are always allowed.
        if PLATFORM_PARAMS.contains(name.as_str()) {
            continue;
        }
        if !supported.contains(name.as_str()) {
            errors.push(format!("{} is not supported by model {}", name, model.id));
            continue;
        }
        if let Some(def) = PARAM_DEFS.get(name.as_str()) {
            if let Some(msg) = validate_param_value(name, value, def) {
                errors.push(msg);
            }
        }
    }
    errors
}

/// Return a map of param name → default value for a model's supported params.
pub fn get_default_parameters(model: &Model) -> HashMap<String, serde_json::Value> {
    let mut defaults = HashMap::new();
    for name in &model.supported_parameters {
        if let Some(def) = PARAM_DEFS.get(name.as_str()) {
            if let Some(ref val) = def.default {
                defaults.insert(name.clone(), val.clone());
            }
        }
    }
    defaults
}

fn validate_param_value(name: &str, value: &serde_json::Value, def: &ParamDef) -> Option<String> {
    match def.param_type.as_str() {
        "number" => {
            let v = value.as_f64()?;
            if let Some(min) = def.min {
                if v < min {
                    return Some(format!("{} must be >= {}", name, min));
                }
            }
            if let Some(max) = def.max {
                if v > max {
                    return Some(format!("{} must be <= {}", name, max));
                }
            }
            None
        }
        "integer" => {
            if let Some(v) = value.as_f64() {
                if v != (v as i64) as f64 {
                    return Some(format!("{} must be an integer", name));
                }
            } else {
                return Some(format!("{} must be an integer", name));
            }
            None
        }
        "boolean" => {
            if !value.is_boolean() {
                return Some(format!("{} must be a boolean", name));
            }
            None
        }
        "string" => {
            if !value.is_string() {
                return Some(format!("{} must be a string", name));
            }
            None
        }
        _ => None,
    }
}
